### ISSUE TITLE
[UX] Path filter/tips is not very useful and unlikely to be found

### ISSUE THREAD

**Issue by bugfolder** [backdrop/backdrop-issues#5912]
I came upon this while working on the [Token Filter](https://backdropcms.org/project/token_filter) module. All text filters have a `tips callback` function that returns either a short or long form of the tips.

The short form is familiar: it's what's shown on the edit page for a text format at admin/config/content/formats/my_format and in the expanded information under a text field that uses that format.

The long form, though, is (AFAIK) only shown on the page filter/tips (and filter/tips/my_filter).

My first question is, how is a visitor (or site architect) going to ever find this page filter/tips? It's not at an admin path, so doesn't get the admin theme, and is visible to anonymous users, so it's presumably intended for the non-admin audience. But how would they ever find it?

There is a theme function, `theme_filter_tips_more_info()`, that returns a link to this page, but I was unable to find any invocations of it (`theme('filter_tips_more_info', ...)`) anywhere in core.

And comparing the short and long forms for the built-in filters (Filtered HTML, Raw HTML) and some others (Markdown), the long and short forms are practically the same. Do we even need both?

And then, the page itself doesn't seem all that useful. Here's what it looks like to an anonymous visitor to https://backdropcms.org/filter/tips:

![compose_tips](https://user-images.githubusercontent.com/29214084/210653283-a8c7d906-fea8-40af-bb7f-a39a86ba7dc7.jpg)

So...

* Do we really need this page (and the associated long form of text filter tips)? 
* If we do, maybe there's a way to get it more easily reachable so that it will serve its intended purpose?
* If it's intended to be used by site contributors, perhaps it could be customizable so that the site architect could add some explanatory text at the top?

**Comment by klonos** [comment]
Possible explanation is that this was used before for people with JS disabled on their browsers and/or that we had some kind of "more details" link in the text format collapsed filter help. I vaguely recall something re us revamping the information for text formats by moving it in a collapsible and collapsed-by-default fieldset, at which point we must have removed the link from but left the page generated for it behind (perhaps an oversight, perhaps intentionally for backwards-compatibility reasons).

I hope this helps @bugfolder

**Comment by klonos** [comment]
...I haven't checked the commit history to confirm that, neither have I span up a D7 site with CKEditor, but in D9/D10 there's an "About text formats" link that points to `/filter/tips`:

![image](https://user-images.githubusercontent.com/2423362/210806265-8faca220-d95f-4ab7-a661-06759954f85a.png)

Here's what that linked page looks like in D10:
![screencapture-demos6-softaculous-Drupals5ir1rgerk-filter-tips-2023-01-05-16_43_33](https://user-images.githubusercontent.com/2423362/210807276-5d84628f-c330-48c5-98ca-ff775f62a31e.png)

**Comment by klonos** [comment]
...initially that seemed like a security concern to me, but I have just checked and confirmed that this page is also accessible by anonymous visitors in D10. I guess because in some sites anonymous users are allowed to add comments or fill in rich text fields in webforms etc.

As I was typing that though, I thought that we already know which text formats anonymous users are allowed to use, so that page should only be showing info for those formats that the anonymous role has access for - not for all of them 🤔

**Comment by bugfolder** [comment]
> that page should only be showing info for those formats that the anonymous role has access for - not for all of them. Right?

Yes, and that seems to be the case for Backdrop's version: anonymous visitors only see the "Filtered HTML" information.

Thanks very much for the background and explanation. I guess there's still the question of whether we want to provide a B-wide way to direct people to this page, or just leave it optional for developers to add a link if they want to (and know about it).

**Comment by kiamlaluno** [comment]
On Drupal 7, `theme_filter_tips_more_info()` is used by [`filter_process_format()`](https://api.drupal.org/api/drupal/modules%21filter%21filter.module/function/filter_process_format/7.x). On Backdrop, [`filter_process_format()`](https://docs.backdropcms.org/api/backdrop/core%21modules%21filter%21filter.module/function/filter_process_format/1) doesn't use it.

The Backdrop function doesn't have the following code.

```php
$element['format']['help'] = array(
  '#type' => 'container',
  '#theme' => 'filter_tips_more_info',
  '#attributes' => array(
    'class' => array(
      'filter-help',
    ),
  ),
  '#weight' => 0,
);
```

**Comment by kiamlaluno** [comment]
I am trying to find an issue that was created to remove that code. So far, I found out that code was previously present on [line 1179](https://github.com/backdrop/backdrop/blame/e312d30583c4149070989e6a493f75909dbda958/core/modules/filter/filter.module#L1179).

**Comment by kiamlaluno** [comment]
The code I shown in my previous comment has been removed in commit [5364a34e632036697cc3a737624210ac5514e2a6](https://github.com/backdrop/backdrop/commit/5364a34e632036697cc3a737624210ac5514e2a6).

From the commit message (*Hide the text format list if only one exists. Remove filter tips.*), I take that was intentional. In that case, `theme_filter_tips_more_info()` should also be removed.

**Comment by kiamlaluno** [comment]
As for why `theme_filter_tips_more_info()` wasn't removed, #1089 (*[UX] Open long filter tips in a dialog and rework the contents*) could explain the reason.

**Comment by bugfolder** [comment]
> From the commit message (_Hide the text format list if only one exists. Remove filter tips._), I take that was intentional. In that case, `theme_filter_tips_more_info()` should also be removed.

If we remove `theme_filter_tips_more_info()`, we're effectively removing the "long" form of filter tips, and so the changes to expunge it should probably be even more extensive.

But before we go down that road: that commit came from @quicksketch, so wondering if @quicksketch could comment on the reasoning for removing it entirely (if recollection serves)?

**Comment by klonos** [comment]
Not sure how to categorize this issue 🤔 ...it's definitely not a feature request. Task? Question?

**Comment by bugfolder** [comment]
Notes from today's Design/UX meeting.

The `filter/tips` path used to be linked from core text filter descriptions using `theme_filter_tips_more_info()`, but that was removed (as @kiamlaluno noted above) because people clicking it would get redirected to a new page (and thereby lose any text they had already entered); so, bad UX.

The reason for having a "more information" link in the first place, rather than just providing full information to begin with, was to keep the length of the format text presented on the node edit page to something reasonably sized under the text being edited. (Each _format_ can contain description from multiple _filters_.) Because the text format information is now wrapped in a collapsed fieldset, that consideration is less important now.

And there was a sense that the presented information under the text box should be complete in and of itself (rather than having to go elsewhere), but that information itself could be presented in a way that was expandable: perhaps using further `'details'` elements within the text filter information, or using a modal dialog for the format information, or for "more information".

For many modules, the short and long tips are the same anyhow, but for some, they are different. (A good sub-task of this issue would be to actually do a survey through core and contrib to see how that breaks down.)

So my understanding of where we want to go with this is:

* We should deprecate the `filter/tips` page
* Consider using the short info for the admin display at admin/config/content/formats/FORMAT, and then reworking the long display to be used in the expanded description under a text field that uses the format, BUT find a way that the expanded description doesn't become impossibly long when it has multiple filters. Perhaps expandable sections (a separate `details` element for each filter in the text format?), or a modal dialog.
* Consider changing `theme_filter_tips_more_info()` so that it produces a modal dialog, rather than a page redirect (if indeed modal dialogs are part of the solution).

Anyone else who was at the meeting, please add details or corrections if I've missed or misstated anything important.

**Comment by klonos** [comment]
> Anyone else who was at the meeting, please add details or corrections if I've missed or misstated anything important.

I think you got it right @bugfolder 👍🏼 ...for those that weren't in the meeting, here's the relevant discussion: https://youtu.be/nBlrOVwaSwY?t=368

**Comment by jenlampton** [comment]
> Consider using the short info for the admin display at admin/config/content/formats/FORMAT, and then reworking the long display to be used in the expanded description under a text field that uses the format, BUT find a way that the expanded description doesn't become impossibly long when it has multiple filters. Perhaps expandable sections (a separate details element for each filter in the text format?), or a modal dialog.

I was actually thinking that we could do away with the short info entirely (but maybe not until Backdrop 2.x) and keep only the long display for use in both locations. It could be tricky to find a way to do that that's backwards compatible, but I think it's worth investigating.

Generally speaking, I really dislike having two different descriptions for one thing (the recent role help text addition, for example) because it gets very confusing for both users and developers to determine when to use which one (and it's often handled incorrectly). The less we can do this, the better.

**Comment by klonos** [comment]
@jenlampton in order to help move this forward, how about we re-introduce the full-length tips in a modal as a first step, and then file a new issue for 2.x, where we completely remove the short description?

...unless we wand to do all that in 1.x.

**Comment by bugfolder** [comment]
To provide some data for this issue, I've looked through core and a bunch of contrib filters to see how they handled the distinction between short and long versions of the filter tips.

I've give some further thoughts in a subsequent comment.

**Core filters**

* `filter_html`: short just lists allowed tags, long is _much_ longer (if no CKEditor and "help" is turned on in filter settings), giving a detailed description of the tags (see screen shot)
* `filter_autop`: short is one sentence, long is two sentences.
* `filter_url`: short and long are the same (one sentence)
* `filter_html_escape`: short and long are the same (one sentence)

**Some contrib filters**

* `markdown`: short is two sentences, long is an itemized list with additional information.
* `php_code`: short is one sentence, long is an itemized list with many items.
* `spamspan`: short and long are the same, four sentences.
* `token_filter`: short is a few sentences and a clickable token browser, long displays the full token tree
* `codefilter`: short = 1 sentence, long = 2 sentences.
* `emptyparagraphkiller`: short = 1 sentence, long = 2 sentences.
* `feef`: short and long are the same (one sentence)
* `iframe`: short and long are the same (one sentence)
* `maildisguise`: short and long are the same (one sentence)
* `mathjax`: short and long are the same (few sentences)
* `filter_parsedown_filter`: short = 1 sentence, long = list of 6 items.
* `pirate`: short and long are the same (one sentence)
* `video_filter`: short = 2 sentences, long = several paragraphs and a list of items.

Screen shot of `filter_html` filter, short and long versions:

![filter tips](https://user-images.githubusercontent.com/29214084/212488608-a721f045-cd97-4dc4-94ee-e64b1b9edab4.png)

**Comment by bugfolder** [comment]
So, based on the above, I think it's safe to say that existing practice is kind of all over the place in terms of short versus long filter tips.

If we could start over, though, here's what I'd prefer (YMMV):

Have only a single version that is used everywhere, but it should include a more/less toggle like https://github.com/backdrop/backdrop-issues/issues/5090 (sorry @klonos ;o)).

I would keep the information in the collapsed fieldset under the edit text box as it is now, and would not further wrap individual filters' tips in a `details` element because that just means a person needs to click-to-expand twice to see anything useful.

And I'm less enamored of using a modal dialog here, because modal dialogs of explanation often pop up covering up the thing that they're explaining. If I were editing text with limited HTML tags, it is nice to see the list of allowed tags underneath (or otherwise adjacent to) the text I"m editing. (A draggable modal would work, though.)

As for how to move forward with this (or something like it), here's one way:

* Finish and merge https://github.com/backdrop/backdrop-issues/issues/5090;
* Edit core tips to include the more/less toggle in the "short" version (possibly having to edit theming functions to properly handle this);
* Deprecate the long version and path `filter/tips`.
